README
NAME
    WWW::Hashbang::Pastebin - command line pastebin

VERSION
    version 0.004

SYNOPSIS
        $ (hostname ; uptime) | curl -F 'p=<-' http://p.hashbang.ca
        http://p.hashbang.ca/f4s2
        $ chromium-browser http://p.hashbang.ca/f4s2+#l2

DESCRIPTION
    This pastebin has no user interface - use "curl" or
    WWW::Hashbang::Pastebin::Client's "p" command to POST paste content.
    Your paste's ID is returned in the "X-Pastebin-ID" header; the URL in
    the "X-Pastebin-URL", as well as the response content.

    Append a plus sign to the URL to get line numbers. Add an anchor like
    "#l1" to jump to the given line number, or click the line number you
    want. The line number for the selected line will be highlighted.

SEE ALSO
    *   <http://sprunge.us>

    *   <http://p.defau.lt>

AVAILABILITY
    The project homepage is <http://p.hashbang.ca>.

    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see
    <https://metacpan.org/module/WWW::Hashbang::Pastebin/>.

SOURCE
    The development version is on github at
    <http://github.com/doherty/WWW-Hashbang-Pastebin> and may be cloned from
    <git://github.com/doherty/WWW-Hashbang-Pastebin.git>

BUGS AND LIMITATIONS
    You can make new bug reports, and view existing ones, through the web
    interface at <https://github.com/doherty/WWW-Hashbang-Pastebin/issues>.

AUTHOR
    Mike Doherty <doherty@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2012 by Mike Doherty.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Changelog for {{$dist->name}}

{{$NEXT}}
 * Show the Perl version on the index page

0.004     2012-04-15
 * Add a Daemon::Control script

0.003     2012-03-31
 * Add missing Template::Toolkit dependency
 * Simplify HTML output mode - now append a + to get line numbering
    * The selected line number will be highlighted
 * Get line numbers right

0.002     2012-03-22
 * Add missing YAML dependency
 * Use Test::DBIx::Class for in-memory SQLite DB
 * Escaping in templates
 * Fix off-by-one error in line numbering

0.001     2012-03-16
 * Initial version


t/etc/schema.pl
use strict;
use warnings;
use DateTime;

{
    schema_class => 'WWW::Hashbang::Pastebin::Schema',
    resultsets => ['Paste'],
    fixture_sets => {
        basic => [
            Paste => [
                [qw( paste_id paste_content paste_deleted paste_date )],
                [1, "test1\nline2", 0, DateTime->now( time_zone => 'UTC' )->subtract( minutes => 5 )],
                [2, 'test2', 1, DateTime->now( time_zone => 'UTC' )->subtract( minutes => 5 )],
                [3, '<omg>', 0, DateTime->now( time_zone => 'UTC' )->subtract( minutes => 4 )],
                [4, "test1\nline2\n\n", 0, DateTime->now( time_zone => 'UTC' )->subtract( minutes => 3 )],
            ]
        ]
    },
};


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Dancer::Test" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "Module::Build" => "0.3601",
    "Test::DBIx::Class" => 0,
    "Test::More" => 0
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "command line pastebin",
  "dist_author" => [
    "Mike Doherty <doherty\@cpan.org>"
  ],
  "dist_name" => "WWW-Hashbang-Pastebin",
  "dist_version" => "0.003",
  "license" => "perl",
  "module_name" => "WWW::Hashbang::Pastebin",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "DBIx::Class::Core" => 0,
    "DBIx::Class::Schema" => 0,
    "Daemon::Control" => 0,
    "Dancer" => 0,
    "Dancer::Plugin::DBIC" => 0,
    "Dancer::Plugin::EscapeHTML" => 0,
    "DateTime" => 0,
    "DateTime::Format::MySQL" => 0,
    "DateTime::Format::SQLite" => 0,
    "Integer::Tiny" => 0,
    "Plack" => 0,
    "SQL::Translator" => "0.11006",
    "Starman" => 0,
    "Template" => 0,
    "Try::Tiny" => 0,
    "YAML" => 0,
    "base" => 0,
    "perl" => "5.014000",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [
    "bin/app.pl",
    "bin/pastebin"
  ]
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/WWW/Hashbang/Pastebin.pm
package WWW::Hashbang::Pastebin;
use strict;
use warnings;
use 5.014000;
use Dancer ':syntax';
use Dancer::Plugin::DBIC qw(schema);
use Dancer::Plugin::EscapeHTML qw(escape_html);
use Integer::Tiny;
use Try::Tiny;
use DateTime;

# ABSTRACT: command line pastebin
# VERSION

=head1 SYNOPSIS

    $ (hostname ; uptime) | curl -F 'p=<-' http://p.hashbang.ca
    http://p.hashbang.ca/f4s2
    $ chromium-browser http://p.hashbang.ca/f4s2+#l2

=for test_synopsis
1;
__END__

=head1 DESCRIPTION

This pastebin has no user interface - use C<curl> or L<WWW::Hashbang::Pastebin::Client>'s
C<p> command to POST paste content. Your paste's ID is returned in the
C<X-Pastebin-ID> header; the URL in the C<X-Pastebin-URL>, as well as the response
content.

Append a plus sign to the URL to get line numbers. Add an anchor like C<#l1> to
jump to the given line number, or click the line number you want. The line number
for the selected line will be highlighted.

=cut

my $mapper = do {
    my $key = config->{key} || join '', ('a'..'z', 0..9);
    $key =~ tr/+//d;
    Integer::Tiny->new($key);
};


get '/' => sub {
    return template 'index', { version => $^V };
};

post '/' => sub {
    my $paste_content = param('p') || param('sprunge');
    my $lang = param('lang');

    unless ($paste_content) {
        status 'bad_request';
        return 'No paste content received';
    }

    my $now = DateTime->now( time_zone => 'UTC' );
    my $row = schema->resultset('Paste')->create({
        paste_content => $paste_content,
        paste_date    => $now,
    });

    my $ext_id  = $mapper->encrypt($row->id);
    my $ext_url = uri_for("/$ext_id");
    $ext_url = "$ext_url?$lang" if $lang;
    debug "Created paste $ext_id: $ext_url";
    headers
        'X-Pastebin-ID'     => $ext_id,
        'X-Pastebin-URL'    => $ext_url;
    return "$ext_url\n";
};

get '/:id' => sub {
    my $ext_id = param('id');
    my $line_nos = ($ext_id =~ s{\+$}{});

    my $int_id = try { $mapper->decrypt( $ext_id ) } || do {
        status 'bad_request';
        return "'$ext_id' is not a valid paste ID";
    };
    debug "paste ID requested: $int_id";

    my $paste = schema->resultset('Paste')->find( $int_id );

    unless ($paste) {
        my $msg = "No such paste as '$ext_id'";
        warning $msg;
        content_type('text/plain');
        status 'not_found';
        return $msg;
    }
    elsif ($paste->deleted) {
        warning "Request was for deleted paste '$ext_id'->'$int_id'";
        status 'gone';
        return "No such paste as '$ext_id'";
    }

    headers 'X-Pastebin-ID' => $ext_id;
    if ($line_nos) {
        my $lines = do {
            open my $in, '<', \$paste->content;
            1 while (<$in>);
            $.; # "return"
        };
        return template paste => { content => $paste->content, lines => [1..$lines] };
    }
    else {
        content_type('text/plain');
        return $paste->content;
    }
};

=head1 SEE ALSO

=over 4

=item * L<http://sprunge.us>

=item * L<http://p.defau.lt>

=back

=cut

true;


lib/WWW/Hashbang/Pastebin/Schema.pm
package WWW::Hashbang::Pastebin::Schema;
use strict;
use warnings;
use Try::Tiny;
use base qw(DBIx::Class::Schema);
# VERSION
# ABSTRACT: Database schema for WWW::Hashbang::Pastebin

__PACKAGE__->load_namespaces();

sub connection {
    my ($self, @info) = @_;
    return $self if !@info && $self->storage;

    my ($storage_class, $args) = ref $self->storage_type ?
    ($self->_normalize_storage_type($self->storage_type),{}) : ($self->storage_type, {});

    $storage_class = 'DBIx::Class::Storage'.$storage_class
        if $storage_class =~ m/^::/;
    try {
        $self->ensure_class_loaded($storage_class);
    }
    catch {
        $self->throw_exception(
            "Unable to load storage class ${storage_class}: $_"
        );
    };
    my $storage = $storage_class->new($self=>$args);
    $storage->connect_info(\@info);
    $self->storage($storage);

    my $dbname = (split /:/, $info[0], 3)[2];
    if ($dbname eq 'dbname=:memory:') {
#        warn "deploying into $dbname";
        $self->deploy();
    }

    return $self;
}

1;


lib/WWW/Hashbang/Pastebin/Schema/Result/Paste.pm
package WWW::Hashbang::Pastebin::Schema::Result::Paste;
use strict;
use warnings;
use base qw/DBIx::Class::Core/;
# VERSION
# ABSTRACT: represents a paste in the pastebin
 
__PACKAGE__->table('paste');
__PACKAGE__->load_components( qw/InflateColumn::DateTime/ );
__PACKAGE__->add_columns(
    'paste_id'   => {
        data_type           => 'bigint',
        is_auto_increment   => 1,
        is_numeric          => 1,
        accessor            => 'id',
    },
    'paste_content' => {
        data_type           => 'text',
        accessor            => 'content',
    },
    'paste_deleted' => {
        data_type           => 'boolean',
        default_value       => 0,
        accessor            => 'deleted',
    },
    'paste_date'    => {
        data_type           => 'datetime',
        accessor            => 'date',
    }
);

__PACKAGE__->set_primary_key('paste_id');
 
1;


Makefile.PL

use strict;
use warnings;

use 5.014000;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "command line pastebin",
  "AUTHOR" => "Mike Doherty <doherty\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Dancer::Test" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "Module::Build" => "0.3601",
    "Test::DBIx::Class" => 0,
    "Test::More" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "WWW-Hashbang-Pastebin",
  "EXE_FILES" => [
    "bin/app.pl",
    "bin/pastebin"
  ],
  "LICENSE" => "perl",
  "NAME" => "WWW::Hashbang::Pastebin",
  "PREREQ_PM" => {
    "DBIx::Class::Core" => 0,
    "DBIx::Class::Schema" => 0,
    "Daemon::Control" => 0,
    "Dancer" => 0,
    "Dancer::Plugin::DBIC" => 0,
    "Dancer::Plugin::EscapeHTML" => 0,
    "DateTime" => 0,
    "DateTime::Format::MySQL" => 0,
    "DateTime::Format::SQLite" => 0,
    "Integer::Tiny" => 0,
    "Plack" => 0,
    "SQL::Translator" => "0.11006",
    "Starman" => 0,
    "Template" => 0,
    "Try::Tiny" => 0,
    "YAML" => 0,
    "base" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.003",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



